src/transformers/models/dac/modeling_dac.py

```diff
@@ -0,0 +1,665 @@
+        super().__init__()
+
+        self.in_proj = weight_norm(nn.Conv1d(config.hidden_size, config.codebook_dim, kernel_size=1))
+        self.out_proj = weight_norm(nn.Conv1d(config.codebook_dim, config.hidden_size, kernel_size=1))
+        self.codebook = nn.Embedding(config.codebook_size, config.codebook_dim)
+
+    def forward(self, hidden_state):
+        """
+        Quantizes the input tensor using a fixed codebook and returns the corresponding codebook vectors.
+
+        Args:
+            hidden_state (`torch.Tensor` of shape `(batch_size, dimension, time_steps)`):
+                Input tensor.
+
+        Returns:
+            quantized_representation (`torch.Tensor`of shape `(batch_size, dimension, time_steps)`).
+                Quantized continuous representation of input, .
+            commitment_loss : torch.Tensor
+                Commitment loss to train encoder to predict vectors closer to codebook entries, shape `(1)`.
+            codebook_loss : torch.Tensor
+                Codebook loss to update the codebook, shape `(1)`.
+            audio_codes (`torch.Tensor` of shape `(batch_size, time_steps)`):
+                Codebook indices for each codebook, quantized discrete representation of input,
+            projected_latents : torch.Tensor of shape `(batch_size, num_codebooks * dimension, time_steps)`
+                Projected latents (continuous representation of input before quantization),
+        """
+
+        projected_latents = self.in_proj(hidden_state)
+        quantized_representation, audio_codes = self.decode_latents(projected_latents)
+
+        commitment_loss = F.mse_loss(projected_latents, quantized_representation.detach(), reduction="none").mean(
+            [1, 2]
+        )
+        codebook_loss = F.mse_loss(quantized_representation, projected_latents.detach(), reduction="none").mean([1, 2])
+
+        quantized_representation = (
+            projected_latents + (quantized_representation - projected_latents).detach()
+        )  # noop in forward pass, straight-through gradient estimator in backward pass
+        quantized_representation = self.out_proj(quantized_representation)
+
+        return quantized_representation, commitment_loss, codebook_loss, audio_codes, projected_latents
+
+    def decode_latents(self, hidden_states):
+        batch_size, hidden_dim, sequence_length = hidden_states.shape
+        encodings = hidden_states.permute(0, 2, 1).reshape(batch_size * sequence_length, hidden_dim)
+        codebook = self.codebook.weight  # codebook: (N x D)
+
+        # L2 normalize encodings and codebook (ViT-VQGAN)
+        encodings = F.normalize(encodings)
+        codebook = F.normalize(codebook)
+
+        # Compute euclidean distance with codebook
+        l2_norm = encodings.pow(2).sum(1, keepdim=True)
+        dist = (l2_norm - 2 * encodings @ codebook.t()) + codebook.pow(2).sum(1, keepdim=True).t()
+
+        indices = (-dist).max(1)[1].reshape(hidden_states.size(0), -1)
```

Comment:
(nit) maybe put the `-` sign on the line before so that we can do:
```python
indices = dist.max(1)[1]
indices = indices.reshape(hidden_states.size(0), -1)
```
Which is marginally cleaner than what's currently written